src/io/obf.rs

```diff
@@ -48,85 +48,67 @@ pub async fn obfuscate<M, SU, SH, ST, R, P>(
         std::fs::create_dir_all(&dir_path).expect("Failed to create directory");
     }
 
-    let public_circuit = &obf_params.public_circuit;
-    let dim = obf_params.params.ring_dimension() as usize;
-    let log_base_q = obf_params.params.modulus_digits();
-    let d = obf_params.d;
     let hash_key = rng.random::<[u8; 32]>();
-    let sampler_uniform = SU::new();
-    let sampler_trapdoor = ST::new(&obf_params.params, obf_params.trapdoor_sigma);
-    let bgg_pubkey_sampler = BGGPublicKeySampler::<_, SH>::new(hash_key, d);
     let public_data = PublicSampledData::<SH>::sample(&obf_params, hash_key);
     log_mem("Sampled public data");
-    let packed_input_size = public_data.packed_input_size;
-    assert_eq!(public_circuit.num_input(), (2 * log_base_q) + (packed_input_size - 1));
-    #[cfg(feature = "debug")]
-    let reveal_plaintexts = [vec![true; packed_input_size - 1], vec![true; 1]].concat();
-    #[cfg(not(feature = "debug"))]
-    let reveal_plaintexts = [vec![true; packed_input_size - 1], vec![false; 1]].concat();
-
-    let pub_key_init =
-        sample_public_key_by_id(&bgg_pubkey_sampler, &obf_params.params, 0, &reveal_plaintexts);
-    log_mem("Sampled pub key init");
 
     let params = Arc::new(obf_params.params);
+    let public_circuit = obf_params.public_circuit;
     let packed_input_size = public_data.packed_input_size;
+    let log_base_q = params.modulus_digits();
+    assert_eq!(public_circuit.num_input(), (2 * log_base_q) + (packed_input_size - 1));
+    let dim = params.ring_dimension() as usize;
+    let d = obf_params.d;
+    let sampler_uniform = SU::new();
+    let sampler_trapdoor = ST::new(&params, obf_params.trapdoor_sigma);
+    let bgg_pubkey_sampler = BGGPublicKeySampler::<_, SH>::new(hash_key, d);
     let packed_output_size = public_data.packed_output_size;
-    let s_bars = sampler_uniform.sample_uniform(&params, 1, d, DistType::BitDist).get_row(0);
-    log_mem("Sampled s_bars");
+
+    /*
+    =============================================================================
+     We reveal all input slots so the evaluator can multiply encodings,
+     but we keep the secret-key slot (last slot) hidden unless we're in debug mode.
+     In the paper, the last slot corresponds to the FHE secret-key vector t.
+
+     Sample the initial public key (level 0) with our reveal flags.
+
+     Sample the initial BGG+ encodings.
+     **NOTE:** the paper treats
+       - c_att  = encoding of evaluator's inputs
+       - c_t    = encoding of the FHE secret-key
+    as two distinct vectors.
+    Here, however, `encodings_init` bundles them into one Vec<bool> of length L:
```

Comment:
The length of `encodings_init` is 1 (for encoding of one) + `packed_input_size` + 1 (encoding of `t`)